uuid-kinds/src/lib.rs

```diff
@@ -20,16 +20,20 @@ use schemars::JsonSchema;
 macro_rules! impl_typed_uuid_kind {
     ($($kind:ident => $tag:literal),* $(,)?) => {
         $(
-            #[cfg_attr(feature = "schemars08", derive(JsonSchema))]
-            pub enum $kind {}
-
-            impl TypedUuidKind for $kind {
-                #[inline]
-                fn tag() -> TypedUuidTag {
-                    // `const` ensures that tags are validated at compile-time.
-                    const TAG: TypedUuidTag = TypedUuidTag::new($tag);
-                    TAG
+            paste::paste! {
+                #[cfg_attr(feature = "schemars08", derive(JsonSchema))]
+                pub enum [< $kind Kind>] {}
+
+                impl TypedUuidKind for [< $kind Kind >] {
+                    #[inline]
+                    fn tag() -> TypedUuidTag {
+                        // `const` ensures that tags are validated at compile-time.
+                        const TAG: TypedUuidTag = TypedUuidTag::new($tag);
+                        TAG
+                    }
                 }
+
+                pub type [< $kind Uuid>] = TypedUuid::<[< $kind Kind >]>;
```

Comment:
I'm also making a type alias for the type, which is where `ZpoolUuid` comes from. IMO this is slightly easier to parse than `TypedUuid::<ZpoolKind>`, but if there are strong opinions against this pattern, I can switch back.

Reply:
There was some discussion about this on the initial PR: https://github.com/oxidecomputer/omicron/pull/4934#issuecomment-1918078400, https://github.com/oxidecomputer/omicron/pull/4934#issuecomment-1918081744. I think I'd vote with that and not use type aliases, although `ZoneUuid` is definitely clearer than `ZoneId` (which was the proposal in that discussion).

Reply:
To be clear I'm weakly voting for `TypedUuid<FooKind>`; if there's disagreement here I'm happy to go with the majority.

Reply:
If I am in the minority, I'm willing to "disagree and commit" to this approach. I totally get that this is a choice which the `newtype_uuid` crate didn't want to have, but `omicron_uuid_kinds` seems like it could be more opinionated.

(it just feels so wordy! `TypedUuid<ZpoolKind>` is like double the characters of `ZpoolUuid`)

Reply:
Having written some code that uses `TypedUuid`, I think I've changed my mind here, and in particular I like the `ZoneUuid` pattern a lot. I'd just want to only use it for this commonly-used type.

If we like it, we should probably commit to this now since the more folks use it the harder it will be to make changes.

It would also be worth updating the readme to include this information.

Reply:
Basically the risk I'm concerned about is that e.g. I just added `TypedUuidRng<T>` in #5488 -- I don't want us to grow another set of aliases for things that aren't commonly used. But I think the UUID types themselves are above the threshold where a type alias is useful, and `TypedUuidRng<T>`, `DbTypedUuid<T>` etc are below that threshold.

Reply:
If @sunshowers is cool with it, I'm cool with it :)

Reply:
> Basically the risk I'm concerned about is that e.g. I just added `TypedUuidRng<T>` in #5488 -- I don't want us to grow another set of aliases for things that aren't commonly used. But I think the UUID types themselves are above the threshold where a type alias is useful, and `TypedUuidRng<T>`, `DbTypedUuid<T>` etc are below that threshold.

This balance makes sense to me. So we create aliases for `TypedUuid<FooKind> -> FooUuid` (and the change from `FooId` to `FooUuid` really helps here I think!), but we _don't_ create aliases for other the other typed UUID supporting cast?

Reply:
Yeah.